include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -0,0 +1,383 @@
+    {
+        return true;
+    }
+};
+
+template<class F>
+struct check_TIR_helper<F, true>
+{
+    template<class MicrofacetCache>
+    static bool __call(NBL_CONST_REF_ARG(F) fresnel, NBL_CONST_REF_ARG(MicrofacetCache) cache)
+    {
+        return ComputeMicrofacetNormal<typename F::scalar_type>::isValidMicrofacet(cache.isTransmission(), cache.getVdotL(), cache.getAbsNdotH(), fresnel.orientedEta);
+    }
+};
+
+template<class F, bool IsBSDF>
+struct getOrientedFresnel;
+
+template<class F>
+struct getOrientedFresnel<F, false>
+{
+    static F __call(NBL_CONST_REF_ARG(F) fresnel, typename F::scalar_type NdotV)
+    {
+        // expect conductor fresnel
+        return fresnel;
+    }
+};
+
+template<class F>
+struct getOrientedFresnel<F, true>
+{
+    using scalar_type = typename F::scalar_type;
+
+    static F __call(NBL_CONST_REF_ARG(F) fresnel, scalar_type NdotV)
+    {
+        fresnel::OrientedEtas<typename F::vector_type> orientedEta = fresnel::OrientedEtas<typename F::vector_type>::create(NdotV, fresnel.orientedEta.value);
+        return F::create(orientedEta);
+    }
+};
+
+template<class F, typename Spectral, bool IsAnisotropic>
+struct CookTorranceParams;
+
+template<class F, typename Spectral>
+struct CookTorranceParams<F,Spectral,false>
+{
+    using scalar_type = typename F::scalar_type;
+    using spectral_type = Spectral;
+
+    scalar_type A;
+    F fresnel;
+};
+
+template<class F, typename Spectral>
+struct CookTorranceParams<F,Spectral,true>
+{
+    using scalar_type = typename F::scalar_type;
+    using spectral_type = Spectral;
+
+    scalar_type ax;
+    scalar_type ay;
+    F fresnel;
+};
+}
+
+// N (NDF), F (fresnel)
+template<class Config, class N, class F NBL_PRIMARY_REQUIRES(config_concepts::MicrofacetConfiguration<Config> && ndf::NDF<N> && fresnel::Fresnel<F>)
+struct SCookTorrance
+{
+    MICROFACET_BXDF_CONFIG_TYPE_ALIASES(Config);
+
+    using this_t = SCookTorrance<Config, N, F>;
+    using quant_type = typename N::quant_type;
+    using ndf_type = N;
+    using fresnel_type = F;
+
+    NBL_CONSTEXPR_STATIC_INLINE bool IsAnisotropic = ndf_type::IsAnisotropic;
+    NBL_CONSTEXPR_STATIC_INLINE bool IsBSDF = ndf_type::IsBSDF;
+
+    using creation_params_type = impl::CookTorranceParams<fresnel_type, spectral_type, IsAnisotropic>;
+
+    template<typename C=bool_constant<!IsAnisotropic> >
+    static enable_if_t<C::value && !IsAnisotropic, this_t> create(NBL_CONST_REF_ARG(creation_params_type) params)
+    {
+        this_t retval;
+        retval.ndf = ndf_type::create(params.A);
+        retval.fresnel = params.fresnel;
+        return retval;
+    }
+    template<typename C=bool_constant<IsAnisotropic> >
+    static enable_if_t<C::value && IsAnisotropic, this_t> create(NBL_CONST_REF_ARG(creation_params_type) params)
+    {
+        this_t retval;
+        retval.ndf = ndf_type::create(params.ax, params.ay);
+        retval.fresnel = params.fresnel;
+        return retval;
+    }
```

Comment:
these creation factories are not needed

Reply:
think about what happens when I add an NDF that has 3 roughness parameters and not 2, or multiple parameters but is still isotropic